### PR TITLE
Clarify how String::leak and into_boxed_str differ 

### PR DESCRIPTION
**Pull request by jonhoo**

### CODE REVIEW

File: library/alloc/src/string.rs [hunk 1]
```diff
@@ -1969,8 +1970,11 @@ impl String {
     ///
     /// It does not reallocate or shrink the `String`,
     /// so the leaked allocation may include unused capacity that is not part
-    /// of the returned slice. If you don't want that, call [`into_boxed_str`],
-    /// and then [`Box::leak`].
+    /// of the returned slice. The advantage of this is that it avoids the potential
+    /// reallocation that [`shrink_to_fit`] may perform.
+    ///
```
Comment (jonhoo) [inline library/alloc/src/string.rs:1975]:
```suggestion
    ///
```

File: library/alloc/src/string.rs [hunk 2]
```diff
@@ -1969,8 +1970,11 @@ impl String {
     ///
     /// It does not reallocate or shrink the `String`,
     /// so the leaked allocation may include unused capacity that is not part
-    /// of the returned slice. If you don't want that, call [`into_boxed_str`],
-    /// and then [`Box::leak`].
+    /// of the returned slice. The advantage of this is that it avoids the potential
+    /// reallocation that [`shrink_to_fit`] may perform.
```
Comment (CryZe) [inline library/alloc/src/string.rs:1974]: Something feels very odd in terms wording here to me. When I read that there's no reallocations, then I wouldn't expect a sentence saying that it's somehow an advantage over some other seemingly unrelated method to leaking.

Comment (jonhoo) [reply library/alloc/src/string.rs]: Yeah, you're not wrong. Gave it another try :+1:

**Comment by Nilstrieb** [comment]
I don't think this change really matches the PR description, it doesn't really make it much clearer what it's useful for. Still, this is a good docs improvement that I'm happy to approve (after you squashed the commits and CI passes).

File: library/alloc/src/string.rs [hunk 3]
```diff
@@ -1940,8 +1940,10 @@ impl String {
 
     /// Converts this `String` into a <code>[Box]<[str]></code>.
     ///
-    /// This will drop any excess capacity.
+    /// Before doing the conversion, this method discards excess capacity like [`shrink_to_fit`].
+    /// Note that the [`shrink_to_fit`] call may reallocate and copy the bytes of the string.
```
Comment (MonterraByte) [inline library/alloc/src/string.rs:1944]: It feels odd to describe a side effect of a `shrink_to_fit` call, considering that `into_boxed_str` isn't previously described as calling `shrink_to_fit` (only that it discards excess capacity, like `shrink_to_fit`).

Comment (qrnch-jan) [reply library/alloc/src/string.rs:1944]: Should probably convey the core point without mentioning implementation details like `shrink_to_fit`.

**Comment by jonhoo** [comment]
> I don't think this change really matches the PR description, it doesn't really make it much clearer what it's useful for. Still, this is a good docs improvement that I'm happy to approve (after you squashed the commits and CI passes).

Heh, apologies for that, I was encouraged (:eyes:) to submit this while doing a live stream, and so my attention to detail is somewhat compromised. Have squashed now and updated the title :+1:

File: library/alloc/src/string.rs (hunk 3 above)
Comment (jonhoo) [reply library/alloc/src/string.rs:1944]: Note that this phrasing was taken from `Vec::into_boxed_slice`, so if this should be changed, it should be changed there as well.

Comment (MonterraByte) [reply library/alloc/src/string.rs:1944]: > Note that this phrasing was taken from `Vec::into_boxed_slice`

That's the first sentence. My comment was about the second one (GitHub's UI does not make this clear, unfortunately).

Comment (jonhoo) [reply library/alloc/src/string.rs]: That's fair — changed the side-effect writing to not talk about `shrink_to_fit` :+1:

**Comment by jonhoo** [comment]
Figured I could pro-actively mark this as a fine to roll up, but I guess not :sweat_smile:

**Comment by Nilstrieb** [comment]
i can do that for you :3, thank you for the PR!
@bors r+ rollup
also greetings to your stream viewers if you're still around, i kindly ask you to meow once or twice for them jon :)

**Pull request merged**